src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,563 @@
 {
-	return clock() != 0;
+	/* only makes sense for setups with an external oscillator */
+	assert(is_external_oscillator());
+
+	m_stream->update();
+
+	return current_clock_state();
+}
+
+void cvsd_device::process_bit(bool bit, bool clock_state)
+{
+	// stub
 }
 
+//-------------------------------------------------
+//  sound_stream_update - handle a stream update
+//-------------------------------------------------
 
-inline int hc55516_device::is_active_clock_transition(int clock_state)
+void cvsd_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
 {
-	return (( m_active_clock_hi && !m_last_clock_state &&  clock_state) ||
-			(!m_active_clock_hi &&  m_last_clock_state && !clock_state));
+	// Stub, just return silence
+	auto &buffer = outputs[0];
+
+	m_samples_generated += buffer.samples();
+	if (m_samples_generated >= SAMPLE_RATE)
+		m_samples_generated -= SAMPLE_RATE;
+	for (int i = 0; i < buffer.samples(); i++)
+		buffer.put(i, stream_buffer::sample_t(0.0));
 }
 
 
-inline int hc55516_device::current_clock_state()
+//#########################################
+//                 HC55516
+//#########################################
+DEFINE_DEVICE_TYPE(HC55516, hc55516_device, "hc55516", "HC-55516")
+
+hc55516_device::hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: cvsd_device(mconfig, HC55516, tag, owner, clock)
+	, m_agc_push_cb(*this)
+	, m_fzq_pull_cb(*this)
+	, m_sylmask(0xfc0)
+	, m_sylshift(6)
+	, m_syladd(0xfc1)
+	, m_intshift(4)
+	, m_sylfilter(0)
+	, m_intfilter(0)
+	, m_agc(true)
+	, m_buffered_fzq(true)
 {
-	return ((uint64_t)m_update_count * clock() * 2 / SAMPLE_RATE) & 0x01;
 }
 
+// overridable type for hc55532 etc
+hc55516_device::hc55516_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+	: cvsd_device(mconfig, type, tag, owner, clock)
+	, m_agc_push_cb(*this)
+	, m_fzq_pull_cb(*this)
+	, m_sylmask(0xf80)
+	, m_sylshift(7)
+	, m_syladd(0xfe1)
+	, m_intshift(5)
+	, m_sylfilter(0)
+	, m_intfilter(0)
+	, m_agc(true)
+	, m_buffered_fzq(true)
+{
+}
 
-void hc55516_device::process_digit()
+//-------------------------------------------------
+//  device_start - device-specific start
+//-------------------------------------------------
+
+void hc55516_device::device_start()
 {
-	double integrator = m_integrator, temp;
+	cvsd_device::device_start();
+	save_item(NAME(m_sylmask));
+	save_item(NAME(m_sylshift));
+	save_item(NAME(m_syladd));
+	save_item(NAME(m_intshift));
+	save_item(NAME(m_sylfilter));
+	save_item(NAME(m_intfilter));
+	save_item(NAME(m_agc));
+	save_item(NAME(m_buffered_fzq));
+
+	/* resolve lines */
+	m_agc_push_cb.resolve();
+	m_fzq_pull_cb.resolve();
+
+	/* variant-specific parameters */
+	m_shiftreg_mask = 0x07;
+	m_active_clock_edge = RISING;
+	m_sylmask = 0xfc0;
+	m_sylshift = 6;
+	m_syladd = 0xfc1;
+	m_intshift = 4;
```

Comment:
These should be arguments to the base device constructor.

Reply:
Fixed.